typing/typemod.ml

```diff
@@ -1163,6 +1167,48 @@ let rec approx_modtype env smty =
       let aliasable = (not (Env.is_functor_arg path env)) in
       Mty_strengthen (mty, path, Aliasability.aliasable aliasable)
 
+and approx_with ~loc env sg constr =
+  let open struct
+    type namespace = Type | Module | Module_type
+  end in
+  let rec remove_from_sg namespace path env sg =
+    List.filter_map (remove_from_sig_item namespace path env sg) sg
+  and remove_from_sig_item namespace path env sg_for_env item =
+    match path, namespace, item with
+    | [s], Type, Sig_type (id, _, _, _)
+        when Ident.name id = s ->
+      None
+    | [s], Module, Sig_module (id, _, _, _, _)
+        when Ident.name id = s ->
+      None
+    | [s], Module_type, Sig_modtype (id, _, _)
+        when Ident.name id = s ->
+      None
+    | s :: path, namespace,
+      Sig_module (id, presence, md, rs, priv)
+        when Ident.name id = s ->
+      let env = Env.add_signature sg_for_env env in
+      let sg = extract_sig env loc md.md_type in
+      let md =
+        { md with
+          md_type = Mty_signature (remove_from_sg namespace path env sg) }
+      in
+      Some (Sig_module (id, presence, md, rs, priv))
+    | _ ->
+      Some item
+  in
+  (* Removing destructively-substituted items approximates [transl_with],
+     as all approximated types are treated as abstract. *)
+  match constr with
+  | Pwith_typesubst (lid, _) ->
+    remove_from_sg Type (Longident.flatten lid.txt) env sg
+  | Pwith_modsubst (lid, _) ->
+    remove_from_sg Module (Longident.flatten lid.txt) env sg
+  | Pwith_modtypesubst (lid, _) ->
+    remove_from_sg Module_type (Longident.flatten lid.txt) env sg
+  | _ ->
```

Comment:
Please make this an or-pattern that lists all the remaining cases, so we're forced to revisit if we add more `with` clauses (which we will, soon, for abstract kinds).